### PR TITLE
Path planner

### PR DESCRIPTION
**Pull request by TrevorDecker**
This fixes several bugs with the path planner and integrates imu/encoders

### CODE REVIEW

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/robots/SimRobot.java [hunk 1]
```diff
@@ -1,22 +1,28 @@
...
+import com.roboclub.robobuggy.ui.ConfigurationPanel;
 import com.roboclub.robobuggy.ui.Gui;
 import com.roboclub.robobuggy.ui.ImuPanel;
 import com.roboclub.robobuggy.ui.MainGuiWindow;
+import com.roboclub.robobuggy.ui.PathPanel;
 import com.roboclub.robobuggy.ui.PoseGraphsPanel;
 import com.roboclub.robobuggy.ui.RobobuggyGUITabs;
 import com.roboclub.robobuggy.ui.RobobuggyJFrame;
 import com.roboclub.robobuggy.ui.SimulationPanel;
 import com.roboclub.robobuggy.ui.VelocityWindow;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/robots/SimRobot.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.ui.VelocityWindow'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2348033630/issues/source?bid=2938665&fileBranchId=3216826#l23)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SimulatedImuNode.java [hunk 2]
```diff
@@ -1,5 +1,6 @@
 package com.roboclub.robobuggy.simulation;
 
+import com.roboclub.robobuggy.messages.IMUAngularPositionMessage;
 import com.roboclub.robobuggy.messages.ImuMeasurement;
 import com.roboclub.robobuggy.messages.MagneticMeasurement;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SimulatedImuNode.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.messages.MagneticMeasurement'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2348033640/issues/source?bid=2938665&fileBranchId=3216826#l5)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/planners/WayPointFollowerPlanner.java [hunk 3]
```diff
@@ -1,12 +1,19 @@
...
+import com.roboclub.robobuggy.nodes.localizers.LocalizerUtil;
 import com.roboclub.robobuggy.ros.NodeChannel;
+import com.roboclub.robobuggy.ui.AnalyticsPanel;
 
 import java.util.ArrayList;
 import java.util.Date;
 
+import org.openstreetmap.gui.jmapviewer.MapMarkerCircle;
+import org.openstreetmap.gui.jmapviewer.MapMarkerDot;
+import org.openstreetmap.gui.jmapviewer.interfaces.MapMarker;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/planners/WayPointFollowerPlanner.java]: Codacy Issue found: [Avoid unused imports such as 'org.openstreetmap.gui.jmapviewer.interfaces.MapMarker'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2348033615/issues/source?bid=2938665&fileBranchId=3216826#l15)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/planners/WayPointFollowerPlanner.java [hunk 4]
```diff
@@ -1,12 +1,19 @@
...
 import com.roboclub.robobuggy.messages.GPSPoseMessage;
 import com.roboclub.robobuggy.messages.GpsMeasurement;
+import com.roboclub.robobuggy.nodes.localizers.LocalizerUtil;
 import com.roboclub.robobuggy.ros.NodeChannel;
+import com.roboclub.robobuggy.ui.AnalyticsPanel;
 
 import java.util.ArrayList;
 import java.util.Date;
 
+import org.openstreetmap.gui.jmapviewer.MapMarkerCircle;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/planners/WayPointFollowerPlanner.java]: Codacy Issue found: [Avoid unused imports such as 'org.openstreetmap.gui.jmapviewer.MapMarkerCircle'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2348033615/issues/source?bid=2938665&fileBranchId=3216826#l13)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java [hunk 5]
```diff
@@ -47,15 +25,27 @@
 	private double theta = 0;
 	private MapPolygonImpl directionLine;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java]: Codacy Issue found: [Avoid unused private fields such as 'directionLine'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2348033651/issues/source?bid=2938665&fileBranchId=3216826#l26)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java [hunk 6]
```diff
@@ -47,15 +25,27 @@
...
+	 * @param destinationPoint the destinationPoint to set
+	 */
+	public synchronized void setDestinationPoint(MapMarkerDot destinationPoint) {
+		this.destinationPoint.setLat(destinationPoint.getLat());
+		this.destinationPoint.setLon(destinationPoint.getLon());
+
+	}
+
+
 	private GPSPoseMessage lastpose;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2348033651/issues/source?bid=2938665&fileBranchId=3216826#l49)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/sensors/HillCrestImuNode.java [hunk 7]
```diff
@@ -30,14 +40,85 @@ public HillCrestImuNode() {
...
 		discover.addListener(this);
 		
+		
 		// TODO Auto-generated constructor stub
 	}
 
+	
+	
 	@Override
 	public void handleRead(Device arg0, HidInMsg arg1, int arg2, long arg3) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/sensors/HillCrestImuNode.java]: Codacy Issue found: [The method handleRead() has an NPath complexity of 256](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2348033619/issues/source?bid=2938665&fileBranchId=3216826#l50)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/sensors/HillCrestImuNode.java [hunk 8]
```diff
@@ -1,26 +1,36 @@
...
+ * Driver for the Hillcrest IMUs
  * @author Trevor Decker
+ * @author Sean Buckley
  *
  */
 public class HillCrestImuNode implements DiscoveryListenerInterface,DeviceListenerInterface, com.roboclub.robobuggy.ros.Node{
 	private Device thisDevice;
-	private Publisher pub = new Publisher(NodeChannel.HILL_CREST_IMU.getMsgPath());
+	private Publisher linearAccPub = new Publisher(NodeChannel.IMU_LINEAR_ACC.getMsgPath());
+	private Publisher linearAccNoGravPub = new Publisher(NodeChannel.IMU_LINEAR_NO_GRAV.getMsgPath());
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/sensors/HillCrestImuNode.java]: Codacy Issue found: [Avoid excessively long variable names like linearAccNoGravPub](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2348033619/issues/source?bid=2938665&fileBranchId=3216826#l29)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/localizers/HighTrustGPSLocalizer.java [hunk 9]
```diff
@@ -42,6 +43,7 @@ public HighTrustGPSLocalizer(){
     	buggyFrameGpsY = 0.0;
```
Comment (msebek) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/localizers/HighTrustGPSLocalizer.java]: Note that subscriber callbacks all run on different threads. you may hit races if you don't plan for this,

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/Map.java [hunk 10]
```diff
@@ -2,89 +2,257 @@
...
-	private BufferedImage map;
-	private ArrayList<LocTuple> locs;
-	private LocTuple imgNorthWest;
-	private LocTuple imgSouthEast;
 
+    private JMapViewerTree mapTree;
+
+    private double mapViewerLat = 40.440138;
+    private double mapViewerLon = -79.945306;
+    private String mapCacheFolderDiskPath = "images/cachedCourseMap";
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/Map.java]: Codacy Issue found: [Avoid excessively long variable names like mapCacheFolderDiskPath](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362883633/issues/source?bid=2938665&fileBranchId=3216826#l46)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/Map.java [hunk 11]
```diff
@@ -2,89 +2,257 @@
...
+        }
+    }
+
+    /**
+     * updates the current arrow displaying on the GUI - shows orientation based on GPS
+     */
+    public void updateArrow() {
+		List<MapPolygon> polygons = getMapTree().getViewer().getMapPolygonList();
+			List<MapMarker> markers = getMapTree().getViewer().getMapMarkerList();
+			if (markers.size() >= 2) {
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/Map.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362883633/issues/source?bid=2938665&fileBranchId=3216826#l178)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java (hunk 5 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java]: Codacy Issue found: [Avoid unused private fields such as 'directionLine'.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362883620/issues/source?bid=2938665&fileBranchId=3216826#l26)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java (hunk 6 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/ui/GpsPanel.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362883620/issues/source?bid=2938665&fileBranchId=3216826#l49)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/robots/SimRobot.java (hunk 1 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/robots/SimRobot.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.ui.VelocityWindow'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362883599/issues/source?bid=2938665&fileBranchId=3216826#l23)

File: real_time/surface_src/java_src/Alice/src/test/java/TestWayPointFollowerPlanner.java [hunk 12]
```diff
@@ -28,24 +30,25 @@ public void test() {
...
 			fail("distance function does not handle some input correctly ");
 		}
 		
-		if(Math.abs(GPSPoseMessage.getDistance(aPose, zeroPose) -  1.802) > .001){
+		if(Math.abs(GPSPoseMessage.getDistance(onePose, aPose) - LocalizerUtil.convertLatToMeters(1.5)) > .001){
 			fail("distance function does not handle some input correctly ");
 		}
 		
-		if(Math.abs(GPSPoseMessage.getDistance(onePose, aPose) - 1.5) > .001){
+		if(Math.abs(GPSPoseMessage.getDistance(onePose, negPose) - LocalizerUtil.convertLonToMeters(2.0)) > .001){
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/test/java/TestWayPointFollowerPlanner.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362883654/issues/source?bid=2938665&fileBranchId=3216826#l41)

File: real_time/surface_src/java_src/Alice/src/test/java/TestWayPointFollowerPlanner.java [hunk 13]
```diff
@@ -28,24 +30,25 @@ public void test() {
 			fail("distance function does not respect idenity");
 		}
 		
-		if(Math.abs(GPSPoseMessage.getDistance(zeroPose, onePose) - 1.0) > .001){
+		if(Math.abs(GPSPoseMessage.getDistance(zeroPose, onePose) - LocalizerUtil.convertLonToMeters(1.0)) > .001){
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/test/java/TestWayPointFollowerPlanner.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362883654/issues/source?bid=2938665&fileBranchId=3216826#l33)

File: real_time/surface_src/java_src/Alice/src/test/java/TestWayPointFollowerPlanner.java [hunk 14]
```diff
@@ -28,24 +30,25 @@ public void test() {
...
 		
-		if(Math.abs(GPSPoseMessage.getDistance(onePose, aPose) - 1.5) > .001){
+		if(Math.abs(GPSPoseMessage.getDistance(onePose, negPose) - LocalizerUtil.convertLonToMeters(2.0)) > .001){
 			fail("distance function does not handle some input correctly ");
 		}
 		
-		if(Math.abs(GPSPoseMessage.getDistance(onePose, negPose) - 2.0) > .001){
+		double dx = LocalizerUtil.convertLonToMeters(1.0);
+		double dy = LocalizerUtil.convertLatToMeters(1.5);
+		if(Math.abs(GPSPoseMessage.getDistance(aPose, zeroPose) -  Math.sqrt(dx*dx+dy*dy)) > .001){
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/test/java/TestWayPointFollowerPlanner.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362883654/issues/source?bid=2938665&fileBranchId=3216826#l47)

File: real_time/surface_src/java_src/Alice/src/test/java/TestWayPointFollowerPlanner.java [hunk 15]
```diff
@@ -28,24 +30,25 @@ public void test() {
 			fail("distance function does not respect idenity");
 		}
 		
-		if(Math.abs(GPSPoseMessage.getDistance(zeroPose, onePose) - 1.0) > .001){
+		if(Math.abs(GPSPoseMessage.getDistance(zeroPose, onePose) - LocalizerUtil.convertLonToMeters(1.0)) > .001){
 			fail("distance function does not handle some input correctly ");
 		}
 		
-		if(Math.abs(GPSPoseMessage.getDistance(aPose, zeroPose) -  1.802) > .001){
+		if(Math.abs(GPSPoseMessage.getDistance(onePose, aPose) - LocalizerUtil.convertLatToMeters(1.5)) > .001){
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/test/java/TestWayPointFollowerPlanner.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2362883654/issues/source?bid=2938665&fileBranchId=3216826#l37)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/robots/TransistorAuton.java [hunk 16]
```diff
@@ -20,6 +22,7 @@
 import com.roboclub.robobuggy.ui.ImuPanel;
 import com.roboclub.robobuggy.ui.ImuVisualWindow;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/robots/TransistorAuton.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.ui.ImuVisualWindow'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2369456077/issues/source?bid=2938665&fileBranchId=3216826#l23)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/robots/TransistorAuton.java [hunk 17]
```diff
@@ -20,6 +22,7 @@
 import com.roboclub.robobuggy.ui.ImuPanel;
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/robots/TransistorAuton.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.ui.ImuPanel'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2369456077/issues/source?bid=2938665&fileBranchId=3216826#l22)

File: real_time/surface_src/java_src/Alice/src/test/java/TestSo2Pose.java [hunk 18]
```diff
@@ -1,105 +1,109 @@
...
 	@Test
 	public void test() {
-		So2Pose A = new So2Pose(0.0, 0.0, 0.0);
-		So2Pose B = new So2Pose(0.0, 1.0, 0.0);
-		So2Pose C = A.mult(B);
-		if(C.getY() != 1.0){		
+		So2Pose a = new So2Pose(0.0, 0.0, 0.0);
+		So2Pose b = new So2Pose(0.0, 1.0, 0.0);
+		So2Pose c = a.mult(b);
+		if(c.getY() != 1.0){
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/test/java/TestSo2Pose.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2369456130/issues/source?bid=2938665&fileBranchId=3216826#l21)

File: real_time/surface_src/java_src/Alice/src/test/java/TestSo2Pose.java [hunk 19]
```diff
@@ -1,105 +1,109 @@
...
 		// test basic compounding of matrices
-		So2Pose A = new So2Pose(0.0,1.0,0.0);
-		So2Pose B = new So2Pose(0.0,2.0,0.0);
-		So2Pose C = A.mult(B);
+		So2Pose a = new So2Pose(0.0,1.0,0.0);
+		So2Pose b = new So2Pose(0.0,2.0,0.0);
+		So2Pose c = a.mult(b);
 		
-		if(C.getY() != 3.0){
+		if(c.getY() != 3.0){
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/test/java/TestSo2Pose.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2369456130/issues/source?bid=2938665&fileBranchId=3216826#l85)

File: real_time/surface_src/java_src/Alice/src/test/java/TestSo2Pose.java [hunk 20]
```diff
@@ -1,105 +1,109 @@
...
 			fail("matrix multiply is not implemented correctly ");
 		}
 		
-		So2Pose D = B.mult(A);
-		if(D.getY() != 3.0){
+		So2Pose d = b.mult(a);
+		if(d.getY() != 3.0){
```
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/test/java/TestSo2Pose.java]: Codacy Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2369456130/issues/source?bid=2938665&fileBranchId=3216826#l90)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SimulatedImuNode.java (hunk 2 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SimulatedImuNode.java]: Codacy Issue found: [Avoid unused imports such as 'com.roboclub.robobuggy.messages.MagneticMeasurement'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2369456086/issues/source?bid=2938665&fileBranchId=3216826#l5)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/planners/WayPointFollowerPlanner.java (hunk 3 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/planners/WayPointFollowerPlanner.java]: Codacy Issue found: [Avoid unused imports such as 'org.openstreetmap.gui.jmapviewer.interfaces.MapMarker'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2369456061/issues/source?bid=2938665&fileBranchId=3216826#l15)

File: real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/planners/WayPointFollowerPlanner.java (hunk 4 above)
Comment (TrevorDecker) [inline real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/planners/WayPointFollowerPlanner.java]: Codacy Issue found: [Avoid unused imports such as 'org.openstreetmap.gui.jmapviewer.MapMarkerCircle'](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2369456061/issues/source?bid=2938665&fileBranchId=3216826#l13)

**Comment by VivaanBahl** [comment]
I am signing off on this, this was tested and has been proven to work sufficiently

**Pull request merged**